src/shared/replication/rules/component.rs

```diff
@@ -0,0 +1,190 @@
+    /// be replicated.
+    ///
+    /// Does not affect initial values or removals.
+    ///
+    /// For example, with a period of 2, any mutation
+    /// will be replicated every second tick.
+    Periodic(u32),
+}
+
+impl SendRate {
+    /// Returns `true` if a mutation for should be replicated on this tick.
+    pub fn send_mutations(self, tick: RepliconTick) -> bool {
+        match self {
+            SendRate::EveryTick => true,
+            SendRate::Once => false,
+            SendRate::Periodic(period) => tick.get() % period == 0,
+        }
+    }
+}
+
+/// Parameters for that can be turned into a component replication rule.
+///
+/// Used for [`IntoComponentRules`] to accept either [`RuleFns`] or a tuple combining
+/// [`RuleFns`] with an associated [`SendRate`].
+///
+/// See [`AppRuleExt::replicate_with`] for more details.
+pub trait IntoComponentRule {
+    /// Turns into a component replication rule and registers its functions in [`ReplicationRegistry`].
+    fn into_rule(self, world: &mut World, registry: &mut ReplicationRegistry) -> ComponentRule;
+}
+
+impl<C: Component<Mutability: MutWrite<C>>> IntoComponentRule for RuleFns<C> {
+    fn into_rule(self, world: &mut World, registry: &mut ReplicationRegistry) -> ComponentRule {
+        let (id, fns_id) = registry.register_rule_fns(world, self);
+        ComponentRule::new(id, fns_id)
+    }
+}
+
+impl<C: Component<Mutability: MutWrite<C>>> IntoComponentRule for (RuleFns<C>, SendRate) {
+    fn into_rule(self, world: &mut World, registry: &mut ReplicationRegistry) -> ComponentRule {
+        let (rule_fns, send_rate) = self;
+        let (id, fns_id) = registry.register_rule_fns(world, rule_fns);
+        ComponentRule {
+            id,
+            fns_id,
+            send_rate,
+        }
+    }
+}
+
+/// Parameters that can be turned into a list of component replication rules.
+///
+/// Implemented for tuples of [`IntoComponentRule`].
+///
+/// See [`AppRuleExt::replicate_with`] for more details.
+pub trait IntoComponentRules {
+    /// Priority when registered with [`AppRuleExt::replicate_with`].
+    ///
+    /// Equals the number of components in a rule.
+    const DEFAULT_PRIORITY: usize;
+
+    /// Turns into a replication rule and registers its functions in [`ReplicationRegistry`].
+    fn into_rules(
+        self,
+        world: &mut World,
+        registry: &mut ReplicationRegistry,
+    ) -> Vec<ComponentRule>;
+}
+
+impl<C: IntoComponentRule> IntoComponentRules for C {
+    const DEFAULT_PRIORITY: usize = 1;
+
+    fn into_rules(
+        self,
+        world: &mut World,
+        registry: &mut ReplicationRegistry,
+    ) -> Vec<ComponentRule> {
+        vec![self.into_rule(world, registry)]
+    }
+}
+
+macro_rules! impl_into_component_rules {
+    ($(($n:tt, $R:ident)),*) => {
+        impl<$($R: IntoComponentRule),*> IntoComponentRules for ($($R,)*) {
+            const DEFAULT_PRIORITY: usize = 0 $(+ { let _ = $n; 1 })*;
```

Comment:
The `let _ = $n;` doesn't seem to do anything here.

Reply:
Yep :slightly_smiling_face: Added a code comment to clarify.